### ISSUE TITLE
python3: clearly define datatypes

### ISSUE THREAD

**Issue by gryphius** [fumail/fuglu#186]
When upgrading to python3 I have had various problems with datatypes. Things that have always been strings in py2 can now be either bytes, unicode strings or even other objects and it's not even guaranteed that a method always returns the same type. For example, the `Suspectfilter.get_field()` filter sometimes returns `Header` objects , `suspect.get_source()` / `set_source()` expect/return bytes. Historically, plugins have always expected to work with strings.

I suggest discuss and define clearly, whether plugins are supposed to work with (unicode) strings or bytes (and do string decoding themselves) , especially around the `fuglu.shared` api. I tend towards the former, eg. let fuglu worry about decoding/encoding and plugins can simply work with unicode strings. But I'm not sure what the consequences are. Maybe we should even provide both interfaces? Eg. "set_source_from_bytes()"  and "set_source_from_string()" etc, like the `email.message` API?

Thoughts?

**Comment by danBLA** [comment]
Yes that's really annoying when moving to python3. In Py2 strings/bytes could be used quite interchangable which now creates lots of problems. Most of the time it's just a bytes <-> str problem moving to Py3. Therefore I created the routines `force_uString`/`force_bString` in `stringencode`. These routines encode/decode if needed (utf-8 by default).

**Thoughts**
Main interface to plugins is `Suspect`. Currently it uses bytes for most of the routines. Which makes sense to me. If we just receive and pass a message we will use bytes (sockets need bytes). So it should be possible to pass a message without modifying it. Therefore we need a bytes version of set_source and get_source. If it has to be decoded/encoded the message might be modified.

We could add `get_source_as_string` and `set_source_from_string` which includes an encoding parameter. It also makes clear the storage in Suspect is not string...

**Comment by danBLA** [comment]
More thoughts, using the Suspect routines defined @gryphius  as example:

I can think of defining return types and create routines to distinguish the input type like:
- Suspect.get_source(maxbytes=None) returns bytes
- Suspect.get_original_source(maxbytes=None) -> bytes
- Suspect.set_source( (bytes) source )
- **new: Suspect.set_source_from_string ( (unicode) string, encoding='utf-8')**
- Suspect.get_headers() -> list containing bytes
- **rename:** Suspect.add_header**string**((unicode) key, (unicode) value, (bool) immediate=False)

or change implementation to make setters flexible:
- Suspect.set_source( (bytes/str/unicode) source, encoding='utf-8' )
-  Suspect.add_header((bytes/str/unicode) key, (bytes/str/unicode) value, (bool) immediate=False, encoding='utf-8')

but clearly define output of getters:
- Suspect.get_headers() -> list containing bytes
- Suspect.get_source(maxbytes=None) returns bytes
- Suspect.get_original_source(maxbytes=None) -> bytes

**Comment by gryphius** [comment]
I would prefer to keep API breakage as small as possible, i.e. the "flexible" approach, so people don't have to rewrite plugins when upgrading. In that respect I'd also prefer to keep get_headers()  (currently a string, not a list) returning str, as this is mostly used to find content in the header section. by returning bytes here we force every caller to decode.   get_(original)_source returning bytes makes sense .

**Comment by danBLA** [comment]
You're right, didn't check the code of `get_headers` carefully enough.
So I would define the interfaces as:

**Suspect**
setters:

- set_source( (bytes/str/unicode) source, encoding='utf-8' )
- add_header((bytes/str/unicode) key, (bytes/str/unicode) value, (bool) immediate=False, encoding='utf-8')

getters:

- get_headers() -> unicode string
- get_source(maxbytes=None) returns bytes
- get_original_source(maxbytes=None) -> bytes

**Suspectfilter**
- strip_text() -> returns unicode string
- get_decoded_textparts() -> returns list of unicode strings

**Suspectfilter.get_field**
- get_field((Suspect) suspect, (bytes/str/unicode) headername) -> return ???
I don't really like it but we might have to return different types based on input, unless someone has a better idea:

if headername points to:

        headers:
            just the headername for normal headers
            mime:headername for attached mime part headers
**-> return unicode string**

        envelope data:
            envelope_from (or from_address)
            envelope_to (or to_address)
            from_domain
            to_domain
            clientip
            clienthostname (fcrdns or 'unknown')
            clienthelo
**-> return unicode string**

        tags
            @tagname
**-> return unicode string**

        body source:
            body:full -> (full source, encoded) 
**-> return bytes**
            body:stripped (or just 'body') : -> returns text/* bodyparts with tags and newlines stripped 
**-> same as get_decoded_textparts stripped and split into a list, so a list of unicode strings**
            body:raw -> decoded raw message body parts 
**-> same as get_decoded_textparts, so unicode string**

**Comment by danBLA** [comment]
I'll implement the interface above...

**Comment by gryphius** [comment]
cool thanks! One suggestion tho: I would consider making Suspectfilter. get_field() *always* returning unicode string, even for the rawbody. Afaik this is only used for pattern matching and templating, so again callers would have to always str() themselves to be sure they can use the output if it's not guaranteed to always be the same type.

**Comment by danBLA** [comment]
ok so last exception remaining for get_field not returning a unicode string would be for
`headername` equals "body:stripped" or "body" where it returns a list of unicode strings. Otherwise `get_field` will always return a unicode string.

**Comment by gryphius** [comment]
oh.. I completely forgot that get_field returned lists of strings so far, not a single string. so I guess it would make sense to keep this behaviour and always return a list of strings

**Comment by danBLA** [comment]
Uups I made a mistake in the `get_field` interface proposal! `get_field` always returns a list (as written in the docstring of the routine :-) )

In Python 2 this is list of str(=bytes). So now it can be made into returning a list of unicode strings
to make it consistent for Python 2 / 3

**Comment by danBLA** [comment]
The implementation of the proposed interface has been merged in #188 ,
so I guess we can close this issue?
